hal/common/mbed_mem_trace.c

```diff
@@ -0,0 +1,119 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include "mbed_mem_trace.h"
+#include "critical.h"
+#include "toolchain.h"
+
+/******************************************************************************
+ * Internal variables, functions and helpers
+ *****************************************************************************/
+
+/* The current tracing mask (see mbed_mem_trace.h for details). */
+static uint8_t mem_trace_mask = MBED_MEM_TRACE_DISABLE;
+/* The callback function that will be called after a traced memory operations finishes. */
+static mbed_mem_trace_cb_t mem_trace_cb;
+/* 'trave_level' guards "trace inside trace" situations (for example, the implementation
+ * of realloc() might call malloc() internally, and since malloc() is also traced, this could
+ * result in two calls to the callback function instead of one. */
+static uint8_t trace_level;
+
+/******************************************************************************
+ * Public interface
+ *****************************************************************************/
+
+void mbed_mem_trace_setup(uint8_t mask, mbed_mem_trace_cb_t cb) {
+    mem_trace_mask = mask;
+    mem_trace_cb = cb;
+}
+
+void* mbed_mem_trace_malloc(void *res, size_t size) {
+    if (mem_trace_cb && (mem_trace_mask & MBED_MEM_TRACE_MALLOC)) {
+        if (core_util_atomic_incr_u8(&trace_level, 1) == 1) {
+            mem_trace_cb(MBED_MEM_TRACE_MALLOC, res, MBED_CALLER_ADDR(), size);
+        }
+        core_util_atomic_decr_u8(&trace_level, 1);
+    }
+    return res;
+}
+
+void* mbed_mem_trace_realloc(void *res, void *ptr, size_t size) {
+    if (mem_trace_cb && (mem_trace_mask & MBED_MEM_TRACE_REALLOC)) {
+        if (core_util_atomic_incr_u8(&trace_level, 1) == 1) {
+            mem_trace_cb(MBED_MEM_TRACE_REALLOC, res, MBED_CALLER_ADDR(), ptr, size);
+        }
+        core_util_atomic_decr_u8(&trace_level, 1);
+    }
+    return res;
+}
+
+void* mbed_mem_trace_calloc(void *res, size_t num, size_t size) {
+    if (mem_trace_cb && (mem_trace_mask & MBED_MEM_TRACE_CALLOC)) {
+        if (core_util_atomic_incr_u8(&trace_level, 1) == 1) {
+            mem_trace_cb(MBED_MEM_TRACE_CALLOC, res, MBED_CALLER_ADDR(), num, size);
+        }
+        core_util_atomic_decr_u8(&trace_level, 1);
+    }
+    return res;
+}
+
+void mbed_mem_trace_free(void *ptr) {
+    if (mem_trace_cb && (mem_trace_mask & MBED_MEM_TRACE_FREE)) {
+        if (core_util_atomic_incr_u8(&trace_level, 1) == 1) {
```

Comment:
This prevents unintended recursion, but can cause problems when multiple threads are involved. If the count is 1 (or higher) and a second thread calls free() the count will not be equal to 1. This would lead to a trace even being dropped for free() and falsely give the appearance of a memory leak.